src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs

```diff
@@ -48,21 +78,36 @@ public static bool EncodeHeader(string name, string value, Span<byte> buffer, ou
                 return false;
             }
 
-            if (!EncodeString(value, buffer.Slice(i), out int valueLength, lowercase: false))
+            if (!EncodeString(value, values, separator, buffer.Slice(i), out int valueLength, lowercase: false))
             {
                 return false;
             }
-
             i += valueLength;
 
             length = i;
             return true;
         }
 
-        private static bool EncodeString(string s, Span<byte> buffer, out int length, bool lowercase)
+        private static bool EncodeStringPart(string s, Span<byte> buffer, ref int i, bool lowercase)
         {
             const int toLowerMask = 0x20;
+            // TODO: use huffman encoding
+            for (int j = 0; j < s.Length; j++)
+            {
+                if (i >= buffer.Length)
+                {
+                    return false;
+                }
 
+                //buffer[i++] = (byte)(s[j] | (lowercase && s[j] >= 'A' && s[j] <= 'Z' ? toLowerMask : 0));
+                buffer[i++] = (byte)(s[j] | (lowercase ? toLowerMask : 0));
```

Comment:
It'd probably be better to have two loops, one if lowercase is true and one if it's false.

Reply:
Going further, I would suggest having two entirely different routines here. Or possibly three.

There are really three cases here:
(1) Encode header name. Always a single value. Checked for validity on construction, so we don't need to check again here (though an assert would be good). Needs to be lower cased, but current lower casing logic is incorrect. See #35164. Note also, ideally this is not the common case because we should optimize known header values, see #31307.
(2) Encode single header value. Always a single value. No lower casing. Needs to be checked for valid ASCII, as we do in HttpConnection but aren't doing here today. See #35165.
(3) Encode multiple header values. Again, needs to be checked for valid ASCII.

Possibly (2) and (3) are the same routine.

Reply:
I'll take a look. I generally preserved existing code but I saw #3516 so I added extra check.